File: backend/src/main/java/com/echonovel/service/AudioService.java
package com.echonovel.service;

import com.echonovel.dto.response.AudioFileResponse;
import org.springframework.web.multipart.MultipartFile;

import java.util.Optional;

/**
 * Service interface for audio operations (TTS + Upload).
 */
public interface AudioService {

    /**
     * Get audio info for a chapter (with access check).
     */
    Optional<AudioFileResponse> getAudioByChapterId(Long chapterId);

    /**
     * Generate TTS audio for a chapter (with access check + caching).
     */
    AudioFileResponse generateTts(Long chapterId, String voice);

    /**
     * Admin upload audio for a chapter.
     */
    AudioFileResponse uploadAudio(Long chapterId, MultipartFile file);
}

File: frontend/src/services/audioService.ts
import api from './api';
import type { ApiResponse, AudioFileResponse } from '../types';

export const audioService = {
  /**
   * Get existing audio for a chapter (returns 404 if none)
   */
  getAudio: async (chapterId: number): Promise<ApiResponse<AudioFileResponse>> => {
    const response = await api.get<ApiResponse<AudioFileResponse>>(`/chapters/${chapterId}/audio`);
    return response.data;
  },

  /**
   * Request TTS audio generation (uses cache if already generated)
   */
  generateTts: async (
    chapterId: number,
    voice: string = 'vi-VN-HoaiMyNeural'
  ): Promise<ApiResponse<AudioFileResponse>> => {
    const response = await api.post<ApiResponse<AudioFileResponse>>(
      `/chapters/${chapterId}/tts?voice=${encodeURIComponent(voice)}`
    );
    return response.data;
  },

  /**
   * Admin upload audio file
   */
  uploadAudio: async (chapterId: number, formData: FormData): Promise<ApiResponse<AudioFileResponse>> => {
    const response = await api.post<ApiResponse<AudioFileResponse>>(
      `/admin/chapters/${chapterId}/audio`,
      formData,
      { headers: { 'Content-Type': 'multipart/form-data' } }
    );
    return response.data;
  },
};
